src/btree/bt_curprev.c

```diff
@@ -385,55 +386,73 @@ __cursor_var_prev(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
         }
 
         /*
-         * If we're at the same slot as the last reference and there's no matching insert list item,
-         * re-use the return information (so encoded items with large repeat counts aren't
-         * repeatedly decoded). Otherwise, unpack the cell and build the return information.
+         * There's no matching insert list item. If we're on the same slot as the last reference,
+         * and the cell is cacheable (it might not be if it's not globally visible), reuse the
+         * previous return data to avoid repeatedly decoding items.
          */
-        if (cbt->cip_saved != cip) {
-            cell = WT_COL_PTR(page, cip);
-            __wt_cell_unpack_kv(session, page->dsk, cell, &unpack);
-            if (unpack.type == WT_CELL_DEL) {
-                if (__wt_cell_rle(&unpack) == 1) {
-                    ++*skippedp;
-                    continue;
-                }
-
-                /*
-                 * There can be huge gaps in the variable-length column-store name space appearing
-                 * as deleted records. If more than one deleted record, do the work of finding the
-                 * next record to return instead of looping through the records.
-                 *
-                 * First, find the largest record in the update list that's smaller than the current
-                 * record.
-                 */
-                ins = __col_insert_search_lt(cbt->ins_head, cbt->recno);
-
-                /*
-                 * Second, for records with RLEs greater than 1, the above call to __col_var_search
-                 * located this record in the page's list of repeating records, and returned the
-                 * starting record. The starting record - 1 is the record to which we could skip, if
-                 * there was no larger record in the update list.
-                 */
-                cbt->recno = rle_start - 1;
-                if (ins != NULL && WT_INSERT_RECNO(ins) > cbt->recno)
-                    cbt->recno = WT_INSERT_RECNO(ins);
+        if (cbt->cip_saved == cip && F_ISSET(cbt, WT_CBT_CACHEABLE_RLE_CELL)) {
+            cbt->iface.value.data = cbt->tmp->data;
```

Comment:
If WT_CBT_CACHEABLE_RLE_CELL is set, can we just return (0) as the cursor data is already referencing the cached value, is that correct? If it's not, when we set the tmp value, we should do a full copy I think.

Reply:
I think so, but I assume there was some reason that logic was there before (it's only been moved around, and now it's skipped sometimes), so I'd hesitate to remove it unless someone knows what that reason was. One possible reason is that because cbt->iface.value is in the public structure the user might have overwritten it; e.g. what happens if you cursor though the table doing get_value and then set_value? It looks like that'll clobber it.

Reply:
I think @quchenhao is right (nice catch!). 

I can't find any place where this code would have any effect, and it's unlikely it's to defend against API buffer changes, we're just shuffling pointers here, we're not actually copying data (as Chenhao pointed out), so if the public buffers get trashed, we're not likely to have anything useful internally.

I'm with David though, I don't see why we aren't getting into trouble with public structures getting overwritten. (Admittedly, it's improbable to set the value and then not use it by calling another WT API, but, yeah, I'm confused.) I think the test that might be worth running before we drop the code, maybe a quick python script:

* create a VLCS with a big RLE,
* next to the middle of it 
* call the python `set_key` and `set_value` functions
* do a next and see what you see

If that finds a problem, we should probably do the full copy.
If that doesn't find a problem, I'd say remove the code (and fix the comment in `include/cursor.h`).

Reply:
Keith and I talked about this and the conclusion is that we're pretty sure that we need to use cbt->tmp and not just re-return what's in cbt->iface.value; I can check as above but I'm pretty sure set_value will change it.
We've also concluded that we need to do the full copy; it's not clear who owns the memory whose pointer we're currently stashing (it's not the public interface and it's not the cursor) and there's no particular reason to think it's going to be stable, especially if the user does a bunch of other stuff between cursor calls. And even if it's stable now, it might not be in a year after further changes to the history store.

Reply:
I've confirmed that we do need to use cbt->tmp; the value in the public cursor changes if you call set_value(), so returning it again can produce wrong results.

Reply:
Digging around some more, it seems that the code that returns the cached value doesn't set WT_CURSTD_VALUE_INT (or clear VALUE_EXT), and didn't before, and definitely should have. However, it also appears that if this is set properly there's logic in the system to call __cursor_localvalue to copy the value when it's no longer safe to point to internal places; e.g. this is called via __wt_session_copy_values from transaction commit. This makes me think that if this infrastructure is functioning properly/as intended, and these flags are manipulated properly, it's not necessary to make a full copy of the value into ->tmp, as long as we have __cursor_localvalue clear WT_CBT_CACHEABLE_RLE_CELL (and clear ->tmp, for good measure), as that'll still make the cursor's public value safe and force the value to be reread on the next visit. Or we could have __cursor_localvalue check for WT_CBT_CACHEABLE_RLE_CELL and do the full copy into ->tmp there. This is a bit of a layering violation, but not a horrible one, and probably preferable to copying the value unconditionally.

Does this seem right? Should we expect this infrastructure to work as intended in the presence of history store stuff? I guess it's also important to notice at this point that the cached value going in ->tmp can't be a history store value, because the CACHEABLE flag is only applied if the value comes from the on-disk RLE cell and is globally visible. So it seems to me that it should be safe, but I'm hardly an expert on this codebase yet...

Reply:
On closer investigation the layer violation is real, not minor; __cursor_localvalue applies to any kind of cursor, and while it's presumably possible to check if it's a btree cursor and then do special things if it is, that seems likely to be ugly.

Furthermore, there's a case that doesn't go through __cursor_localvalue that can invalidate the cached value, namely taking one of the other paths through __cursor_var_next/prev -- it is ok to invalidate the value currently in the public cursor in the course of producing a new one. In principle these other paths could clear WT_CBT_CACHEABLE_RLE_CELL if they cause it to be invalidated, but the code is not currently structured in a way that makes this very practical.

So I think the full copy is, unfortunately, necessary.

Reply:
I've pushed the code with the full copy. (and merged again)